design/animation.md

```diff
@@ -0,0 +1,360 @@
+	1. No easing - Setting fields that cannot be interpolated should also be supported (string, boolean, etc.).
+4. Chaining - Begin an animation immediately after another finishes, or on some other event (behavior tie-in?)
+5. Multiple targets - Animate multiple actors/assets, and multiple properties, in lock-step (equivalent to Unity's
+	animation curves)
+6. Blending - Allow multiple animations to run on the same target simultaneously, and blend the computed destinations
+	into one, proportional to a weight property.
+
+To address this need, I propose the following API:
+
+API Design
+-----------
+
+Animations are assets. They are created explicitly by a function on asset containers (
+`AssetContainer.createAnimation`). Native animations from library or glTF actors do not create these assets, instead
+creating animation instances only (see [Instances](#instances)).
+
+When created directly, animations are created as an array of "tracks", each composed of:
+1. A generic path to a field that can be animated (see [Paths](#paths)).
+2. An array of keyframes, each composed of:
+	1. A time value.
+	2. A valid value for the targeted path. Can also be a reference to another actor/material, which will be evaluated
+		at runtime. See [References](#references).
+	3. An optional bezier easing function.
+
+```ts
+type Animatible = Actor | Material;
+type AnimatibleClassName = 'actor' | 'material';
+type AnimatibleLike = ActorLike | MaterialLike;
+type AnimationProp = Vector2 | Vector3 | Quaternion | Color3 | Color4 | number | string | boolean;
+type EasingFunction = [number, number, number, number];
+
+class AssetContainer {
+	public createAnimation(name: string, options: {
+		targets: { [name: string]: AnimatibleClassName },
+		tracks: Track<AnimationProp>[]
+	}): Animation;
+}
+class Animation {
+	id: Readonly<string>;
+	duration: Readonly<number>;
+	targets: { [name: string]: AnimatibleClassName };
+}
+
+type Track<T extends AnimationProp> = {
+	target: TargetPath<T>;
+	keyframes: Keyframe<T>[];
+}
+
+type Keyframe<T extends AnimationProp> = {
+	time: number;
+	value: T | TargetPath<T>;
+	relative?: boolean;
+	easing?: EasingFunction;
+}
+```
+
+### Paths
+
+Paths are references to properties on an object that can be animated (Actor, Material, etc.). Paths do not reference
+properties on a *particular* object, but instead reference a generic object of a given type by index. When bound,
+the animation will be given the list of objects which it will be modifying, and these path indices refer to that
+argument. Over the wire these are represented as strings, but are easily generated in the API with helper functions.
+
+```ts
+class TargetPath<T> {
+	private parent: TargetPath<any>;
+	constructor(name: string, children: Map<string, TargetPath<any>>) {
+		for (const [k,v] of children) {
+			this[k] = v;
+			v.parent = this;
+		}
+	}
+	public toJSON() {
+		// convert tree into string
+	}
+}
+function AnimateActor(name: string) {
```

Comment:
The name `AnimateActor` seems overly specific to one use case. Conceivably, this could be used for things other than animations. What about `ReflectActor`.

Reply:
I suppose the animation-specific part of the function is that it doesn't expose non-animatable fields.

Reply:
I think we should expose non-animatable fields here so that they can be references as destination values in keyframes.